File: project2/src/game_manager/players.rs
use crate::card::Card;

/// Keep track of the cards a player has
/// Keep track of the chips (money)
/// Keep track of if the player has folded
/// Keep track of how much the player has bet in the current round
pub struct Player {
    cards: (Card, Card),
    chips: u32,
    folded: bool,
    bet: u32,
}

impl Player {
    pub fn new(cards: (Card, Card)) -> Player {
        let chips = 100; // TODO: Implement hand value
        Player {
            cards,
            chips,
            folded: false,
            bet: 0,
        }
    }

    pub fn bet(&mut self, amount: u32) -> Result<(), String> {
        if self.chips < amount {
            return Err("Not enough chips".to_string());
        }
        self.chips -= amount;
        Ok(())
    }
}

impl std::fmt::Display for Player {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "Cards: {}, {} ", self.cards.0, self.cards.1)?;
        write!(f, "Chips: {}", self.chips)
    }
}

/// The players struct contains a vector of players
/// Keep track of the dealer
/// Keep track of the highest bet
pub struct Players {
    pub players: Vec<Player>,
    // TODO: Impleemnt if go bust, remember that impacts the choice of the dealer
}

impl Players {
    pub fn len(&self) -> usize {
        self.players.len()
    }

    pub fn iter(&self) -> std::slice::Iter<Player> {
        self.players.iter()
    }
}

impl Players {
    fn get_number_of_folded(&self) -> usize {
        // Note: This is more expensive than a simple counter. However, the number of players is
        // small so this is not an issue, and it is more readable.
        self.players.iter().filter(|player| player.folded).count()
    }

    pub fn get_number_of_active_players(&self) -> usize {
        // Number of players - number of folded players
        self.players.len() - self.get_number_of_folded()
    }

    pub fn has_folded(&self, player: usize) -> bool {
        self.players[player].folded
    }

    pub fn fold(&mut self, player: usize) {
        self.players[player].folded = true;
    }

    pub fn get_bet(&self, player: usize) -> u32 {
        self.players[player].bet
    }

    pub fn place_bet(&mut self, player: usize, bet: u32) -> Result<(), String> {
        self.players[player].bet += bet;
        self.players[player].bet(bet)
    }
}

File: project2/src/game_manager/players.py
import random
from typing import List
from src.game_state.player_state import PublicPlayerState, PrivatePlayerState
from src.poker_oracle.deck import Deck, Card
import copy


class _Player(PrivatePlayerState):
    def __init__(self, ai: bool = False):
        super().__init__()
        self.ai = ai


class Players:
    def __init__(self, num_players: int, num_ai: int):
        self.players = []
        self.players.extend([_Player() for _ in range(num_players)])
        self.players.extend([_Player(ai=True) for _ in range(num_ai)])
        random.shuffle(self.players)

    def get(self, index):
        return self.players[index]

    def reset_round(self, deck: Deck) -> Deck:
        # Deal cards to players
        for player in self.players:
            first_card = deck.pop()
            second_card = deck.pop()

            if not isinstance(first_card, Card):
                raise TypeError(
                    "first_card must be a Card, not {}".format(
                        type(first_card))
                )
            if not isinstance(second_card, Card):
                raise TypeError(
                    "second_card must be a Card, not {}".format(
                        type(second_card))
                )

            player.reset_round((first_card, second_card))
        return deck

    def winner(self, player: int, amount: int):
        self.players[player].chips += amount

    def is_ai(self, player: int) -> bool:
        return self.players[player].ai

    def get_private_player_states(self) -> List[PrivatePlayerState]:
        return self.players

    def get_public_player_states(self) -> List[PublicPlayerState]:
        return [player.to_public() for player in self.players]

    def __len__(self):
        return len(self.players)

    def __iter__(self):
        return iter(self.players)

    def get_number_of_folded(self) -> int:
        # Note: This is more expensive than a simple counter.
        # However, the number of players is small so this is not an issue,
        # and it is more readable.
        # TODO: Consider using a counter instead.
        return sum(player.folded for player in self.players)

    def get_number_of_bust(self) -> int:
        return sum(player.bust for player in self.players)

    def get_number_of_active_players(self) -> int:
        return (
            len(self.players) - self.get_number_of_folded() -
            self.get_number_of_bust()
        )

    def get_active_players(self) -> List[int]:
        return [i for i, player in enumerate(self.players) if self.is_active(i)]

    def get_active_player(self) -> int:
        assert len(self.get_active_players()) == 1
        return self.get_active_players()[0]

    def has_folded(self, player: int) -> bool:
        return self.players[player].folded

    def is_bust(self, player: int) -> bool:
        return self.players[player].bust

    def fold(self, player: int):
        self.players[player].folded = True

    def get_bet(self, player: int) -> int:
        return self.players[player].round_bet

    def action(self, player, action) -> bool:
        return self.players[player].action(action, action.amount)

    def is_active(self, player: int) -> bool:
        return not self.players[player].folded and not self.players[player].bust

    def get_cards(self, player: int) -> List[Card]:
        return self.players[player].cards
